hooks/tk-multi-publish2/basic/collector.py

```diff
@@ -0,0 +1,291 @@
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+
+class MariSessionCollector(HookBaseClass):
+    """
+    Collector that operates on the mari session. Should inherit from the basic
+    collector hook.
+    """
+
+    @property
+    def settings(self):
+        """
+        Dictionary defining the settings that this collector expects to receive
+        through the settings parameter in the process_current_session and
+        process_file methods.
+
+        A dictionary on the following form::
+
+            {
+                "Settings Name": {
+                    "type": "settings_type",
+                    "default": "default_value",
+                    "description": "One line description of the setting"
+            }
+
+        The type string should be one of the data types that toolkit accepts as
+        part of its environment configuration.
+        """
+
+        # grab any base class settings
+        collector_settings = super(MariSessionCollector, self).settings or {}
+
+        # settings specific to this collector
+        mari_session_settings = {
+            "Work Template": {
+                "type": "template",
+                "default": None,
+                "description": "Template path for artist work files. Should "
+                               "correspond to a template defined in "
+                               "templates.yml. If configured, is made available"
+                               "to publish plugins via the collected item's "
+                               "properties. ",
+            },
+        }
+
+        # update the base settings with these settings
+        collector_settings.update(mari_session_settings)
+
+        return collector_settings
+
+    def process_current_session(self, settings, parent_item):
+        """
+        Analyzes the current session open in Mari and parents a subtree of
+        items under the parent_item passed in.
+
+        :param dict settings: Configured settings for this collector
+        :param parent_item: Root item instance
+        """
+
+        # create an item representing the current mari session
+        channel_item, layer_item = self.collect_current_mari_session(settings, parent_item)
+
+        self._collect_files(settings, channel_item, layer_item)
+
+    def collect_current_mari_session(self, settings, parent_item):
+        """
+        Creates an item that represents the current mari session.
+
+        :param dict settings: Configured settings for this collector
+        :param parent_item: Parent Item instance
+        :returns: Items of type mari.channels and mari.layers
+        """
+
+        if not mari.projects.current():
+            self.logger.warning("You must be in an open Mari project to be able to publish!")
+
+        publisher = self.parent
+
+        # create the channel item for the publish hierarchy
+        channel_item = parent_item.create_item(
+            "mari.channels",
+            "Publish flattened channels",

```

Comment:
We tend to think of these items as the nouns and the publish plugins as the verbs. Maybe we should rename this? So we have "what the thing is here" (file name, or "current session") and the next argument is the "type of thing".